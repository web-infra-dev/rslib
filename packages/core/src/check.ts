import type { RsbuildConfig, RsbuildPlugin } from '@rsbuild/core';
import type { TsconfigCompilerOptions } from './types';
import { color } from './utils/helper';
import { logger } from './utils/logger';

type PluginReactOptions = {
  tsconfigCompilerOptions?: TsconfigCompilerOptions;
};

const mapTsconfigJsxToSwcJsx = (jsx: string | undefined): string | null => {
  if (jsx === undefined) {
    // 'preserve' is the default value of tsconfig.compilerOptions.jsx
    return null;
  }

  // Calculate a corresponding SWC JSX config if tsconfig.compilerOptions.jsx is set to React related option.
  // Return `null` stands for no need to check.
  switch (jsx) {
    case 'react-jsx':
    case 'react-jsxdev':
      return 'automatic';
    case 'react':
      return 'classic';
    case 'preserve':
    case 'react-native':
      // SWC JSX does not support `preserve` as of now.
      return null;
    default:
      return null;
  }
};

const checkJsx = ({
  tsconfigCompilerOptions,
}: PluginReactOptions): RsbuildPlugin => ({
  name: 'rsbuild:lib-check',
  setup(api) {
    api.onBeforeEnvironmentCompile(({ environment }) => {
      const config = api.getNormalizedConfig({
        environment: environment.name,
      });
      const swc = config.tools.swc;
      const tsconfigJsx = tsconfigCompilerOptions?.jsx;
      if (swc && !Array.isArray(swc) && typeof swc !== 'function') {
        const swcReactRuntime = swc?.jsc?.transform?.react?.runtime || null;
        const mapped = mapTsconfigJsxToSwcJsx(tsconfigJsx);
        if (mapped !== swcReactRuntime) {
          logger.warn(
            `JSX runtime is set to ${color.green(`${JSON.stringify(swcReactRuntime)}`)} in SWC, but got ${color.green(`${JSON.stringify(tsconfigJsx)}`)} in tsconfig.json. This may cause unexpected behavior, considering aligning them.`,
          );
        }
      }
    });
  },
});

export const composeCheckConfig = (
  compilerOptions: TsconfigCompilerOptions,
): RsbuildConfig => {
  return { plugins: [checkJsx({ tsconfigCompilerOptions: compilerOptions })] };
};
