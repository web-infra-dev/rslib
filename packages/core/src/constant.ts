export const DEFAULT_CONFIG_NAME = 'rslib.config';

export const DEFAULT_CONFIG_EXTENSIONS = [
  '.js',
  '.ts',
  '.mjs',
  '.mts',
  '.cjs',
  '.cts',
] as const;

export const SWC_HELPERS = '@swc/helpers';

export const JS_EXTENSIONS: string[] = [
  'js',
  'mjs',
  'jsx',
  'ts',
  'mts',
  'tsx',
  'cjs',
  'cjsx',
  'mjsx',
  'mtsx',
  'cts',
  'ctsx',
] as const;

export const CSS_EXTENSIONS: string[] = [
  'css',
  'sass',
  'scss',
  'less',
] as const;

export const ENTRY_EXTENSIONS: string[] = [
  ...JS_EXTENSIONS,
  ...CSS_EXTENSIONS,
] as const;

export const JS_EXTENSIONS_PATTERN: RegExp = new RegExp(
  `\\.(${JS_EXTENSIONS.join('|')})$`,
);

export const CSS_EXTENSIONS_PATTERN: RegExp = new RegExp(
  `\\.(${CSS_EXTENSIONS.join('|')})$`,
);

export const ENTRY_EXTENSIONS_PATTERN: RegExp = new RegExp(
  `\\.(${ENTRY_EXTENSIONS.join('|')})$`,
);
