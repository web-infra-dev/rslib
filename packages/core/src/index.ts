export { prepareCli } from './cli/prepare';
export { runCli } from './cli/commands';
export { defineConfig, loadConfig } from './config';
export { build } from './build';
export { logger } from './utils/logger';

export const version: string = RSLIB_VERSION;
