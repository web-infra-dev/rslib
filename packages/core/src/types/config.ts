import type {
  EnvironmentConfig,
  OutputConfig,
  RsbuildConfig,
  Rspack,
} from '@rsbuild/core';
import type { GetAsyncFunctionFromUnion } from './utils';

export type Format = 'esm' | 'cjs' | 'umd' | 'mf' | 'iife';

export type FixedEcmaVersions =
  | 'es5'
  | 'es6'
  | 'es2015'
  | 'es2016'
  | 'es2017'
  | 'es2018'
  | 'es2019'
  | 'es2020'
  | 'es2021'
  | 'es2022'
  | 'es2023';
export type LatestEcmaVersions = 'es2024' | 'esnext';
export type EcmaScriptVersion = FixedEcmaVersions | LatestEcmaVersions;

export type RsbuildConfigWithLibInfo = {
  id?: string;
  format: Format;
  config: EnvironmentConfig;
};

export type RsbuildConfigEntry = NonNullable<
  NonNullable<EnvironmentConfig['source']>['entry']
>;
export type RsbuildConfigEntryItem = RsbuildConfigEntry[string];
export type RspackResolver = GetAsyncFunctionFromUnion<
  ReturnType<NonNullable<Rspack.ExternalItemFunctionData['getResolve']>>
>;

export type RsbuildConfigOutputTarget = NonNullable<
  EnvironmentConfig['output']
>['target'];

export type Syntax =
  // ECMAScript versions as an common used addition to browserslist query
  | EcmaScriptVersion
  // Support inline browserslist query, like defined in package.json
  | string[];

export type Dts =
  | {
      /**
       * Whether to bundle the declaration files.
       * @defaultValue `false`
       * @see {@link https://rslib.rs/config/lib/dts#dtsbundle}
       */
      bundle?: boolean;
      /**
       * The output directory of declaration files.
       * @defaultValue {@link https://rslib.rs/config/lib/dts#default-value}
       * @see {@link https://rslib.rs/config/lib/dts#dtsdistpath}
       */
      distPath?: string;
      /**
       * Whether to generate declaration files with building the project references.
       * @defaultValue `false`
       * @see {@link https://rslib.rs/config/lib/dts#dtsbuild}
       */
      build?: boolean;
      /**
       * Whether to abort the build process when an error occurs during declaration files generation.
       * @defaultValue `true`
       * @see {@link https://rslib.rs/config/lib/dts#dtsabortonerror}
       */
      abortOnError?: boolean;
      /**
       * Whether to automatically set the declaration file extension based on the {@link format} option.
       * @defaultValue `false`
       * @see {@link https://rslib.rs/config/lib/dts#dtsautoextension}
       */
      autoExtension?: boolean;
    }
  | boolean;

export type AutoExternal =
  | boolean
  | {
      /**
       * Whether to automatically externalize dependencies of type `dependencies`.
       * @defaultValue `true`
       * @see {@link https://rslib.rs/config/lib/auto-external#autoexternaldependencies}
       */
      dependencies?: boolean;
      /**
       * Whether to automatically externalize dependencies of type `optionalDependencies`.
       * @defaultValue `true`
       * @see {@link https://rslib.rs/config/lib/auto-external#autoexternaloptionaldependencies}
       */
      optionalDependencies?: boolean;
      /**
       * Whether to automatically externalize dependencies of type `peerDependencies`.
       * @defaultValue `true`
       * @see {@link https://rslib.rs/config/lib/auto-external#autoexternalpeerdependencies}
       */
      peerDependencies?: boolean;
      /**
       * Whether to automatically externalize dependencies of type `devDependencies`.
       * @defaultValue `false`
       * @see {@link https://rslib.rs/config/lib/auto-external#autoexternaldevdependencies}
       */
      devDependencies?: boolean;
    };

export type BannerAndFooter = {
  js?: string;
  css?: string;
  dts?: string;
};

export type Shims = {
  /**
   * Configure the shims for CommonJS output.
   * @see {@link https://rslib.rs/config/lib/shims#shimscjs}
   */
  cjs?: {
    /**
     * Whether to inject shims for the `import.meta.url` in CommonJS output.
     * @defaultValue `true`
     * @see {@link https://rslib.rs/config/lib/shims#shimscjsimportmetaurl}
     */
    'import.meta.url'?: boolean;
  };
  /**
   * Configure the shims for ESM output.
   * @see {@link https://rslib.rs/config/lib/shims#shimsesm}
   */
  esm?: {
    /**
     * Whether to inject shims for the global `__filename` of CommonJS in ESM output.
     * @defaultValue `false`
     * @see {@link https://rslib.rs/config/lib/shims#shimsesm__filename}
     */
    __filename?: boolean;
    /**
     * Whether to inject shims for the global `__dirname` of CommonJS in ESM output.
     * @defaultValue `false`
     * @see {@link https://rslib.rs/config/lib/shims#shimsesm__dirname}
     */
    __dirname?: boolean;
    /**
     * Whether to inject shims for the global `require` of CommonJS in ESM output.
     * @defaultValue `false`
     * @see {@link https://rslib.rs/config/lib/shims#shimsesmrequire}
     */
    require?: boolean;
  };
};

export type JsRedirect = {
  /**
   * Whether to automatically redirect the import paths of JavaScript output files.
   * @defaultValue `true`
   */
  path?: boolean;
  /**
   * Whether to automatically redirect the file extension to import paths based on the JavaScript output files.
   * @defaultValue `true`
   */
  extension?: boolean;
};

export type StyleRedirect = {
  /**
   * Whether to automatically redirect the import paths of style output files.
   * @defaultValue `true`
   */
  path?: boolean;
  /**
   * Whether to automatically redirect the file extension to import paths based on the style output files.
   * @defaultValue `true`
   */
  extension?: boolean;
};

export type DtsRedirect = {
  /**
   * Whether to automatically redirect the import paths of TypeScript declaration output files.
   * @defaultValue `true`
   */
  path?: boolean;
  /**
   * Whether to automatically redirect the file extension to import paths based on the TypeScript declaration output files.
   * @defaultValue `false`
   */
  extension?: boolean;
};

export type Redirect = {
  /** Controls the redirect of the import paths of output JavaScript files. */
  js?: JsRedirect;
  /** Controls the redirect of the import paths of output style files. */
  style?: StyleRedirect;
  /** Controls the redirect of the import paths of output asset files. */
  asset?: boolean;
  /** Controls the redirect of the import paths of output TypeScript declaration files. */
  dts?: DtsRedirect;
};

export interface LibConfig extends EnvironmentConfig {
  /**
   * The unique identifier of the library.
   * @defaultValue `undefined`
   * @see {@link https://rslib.rs/config/lib/id}
   */
  id?: string;
  /**
   * Output format for the generated JavaScript files.
   * @defaultValue `'esm'`
   * @see {@link https://rslib.rs/config/lib/format}
   */
  format?: Format;
  /**
   * Whether to bundle the library.
   * @defaultValue `true`
   * @see {@link https://rslib.rs/config/lib/bundle}
   */
  bundle?: boolean;
  /**
   * Whether to automatically set the file extension based on {@link format} option in the JavaScript output files.
   * @defaultValue `true`
   * @see {@link https://rslib.rs/config/lib/auto-extension}
   */
  autoExtension?: boolean;
  /**
   * Whether to automatically externalize dependencies of different dependency types and do not bundle them.
   * @defaultValue `true` when {@link format} is `cjs` or `esm`, `false` when {@link format} is `umd` or `mf`.
   * @see {@link https://rslib.rs/config/lib/auto-external}
   */
  autoExternal?: AutoExternal;
  /**
   * Configure the redirect of the import paths, applicable {@link bundle} is set to `false`.
   * @defaultValue `{}`
   * @see {@link https://rslib.rs/config/lib/redirect}
   */
  redirect?: Redirect;
  /**
   * Configure the syntax to which JavaScript and CSS will be downgraded.
   * @defaultValue `'esnext'`
   * @see {@link https://rslib.rs/config/lib/syntax}
   */
  syntax?: Syntax;
  /**
   * Whether to import SWC helper functions from `@swc/helpers` instead of inlining them.
   * @defaultValue `false`
   * @see {@link https://rslib.rs/config/lib/external-helpers}
   */
  externalHelpers?: boolean;
  /**
   * Inject content into the top of each JavaScript, CSS or declaration file.
   * @defaultValue `{}`
   * @see {@link https://rslib.rs/config/lib/banner}
   */
  banner?: BannerAndFooter;
  /**
   * Inject content into the bottom of each JavaScript, CSS or declaration file.
   * @defaultValue `{}`
   * @see {@link https://rslib.rs/config/lib/footer}
   */
  footer?: BannerAndFooter;
  /**
   * Configure the shims for CommonJS and ESM output.
   *
   * @defaultValue
   * ```js
   * const defaultShims = {
   *   cjs: {
   *     'import.meta.url': true,
   *   },
   *   esm: {
   *     __filename: false,
   *     __dirname: false,
   *     require: false,
   *   },
   * };
   * ```
   * @see {@link https://rslib.rs/config/lib/shims}
   */
  shims?: Shims;
  /**
   * Configure the generation of the TypeScript declaration files.
   * @defaultValue `false`
   * @see {@link https://rslib.rs/config/lib/dts}
   */
  dts?: Dts;
  /**
   * The export name of the UMD bundle.
   * @defaultValue `undefined`
   * @see {@link https://rslib.rs/config/lib/umd-name}
   */
  umdName?: Rspack.LibraryName;
  /**
   * The base directory of the output files.
   * @defaultValue `undefined`
   * @see {@link https://rslib.rs/config/lib/out-base}
   */
  outBase?: string;
  /**
   * @inheritdoc
   */
  output?: RslibOutputConfig;
}

export type LibOnlyConfig = Omit<LibConfig, keyof EnvironmentConfig>;

interface RslibOutputConfig extends OutputConfig {
  /**
   * @override
   * @default 'node'
   */
  target?: OutputConfig['target'];
  /**
   * @override
   * @default false
   */
  filenameHash?: OutputConfig['filenameHash'];
  /**
   * @override
   * When minify is not specified, Rslib will use a sane default for minify options.
   * The default options will only perform dead code elimination and unused code elimination.
   *
   * @see {@link https://rslib.rs/config/rsbuild/output#outputminify}
   */
  minify?: OutputConfig['minify'];
}

export interface RslibConfig extends RsbuildConfig {
  lib: LibConfig[];
  /**
   * @inheritdoc
   */
  output?: RslibOutputConfig;
}

export type ConfigParams = {
  env: string;
  command: string;
  envMode?: string;
};

export type RslibConfigSyncFn = (env: ConfigParams) => RslibConfig;
export type RslibConfigAsyncFn = (env: ConfigParams) => Promise<RslibConfig>;
export type RslibConfigExport =
  | RslibConfig
  | RslibConfigSyncFn
  | RslibConfigAsyncFn;
