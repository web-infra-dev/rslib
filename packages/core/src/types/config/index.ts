import type { RsbuildConfig } from '@rsbuild/core';
import type { PluginDtsOptions } from 'rsbuild-plugin-dts';

export type Format = 'esm' | 'cjs' | 'umd';

export type FixedEcmaVersions =
  | 'es5'
  | 'es6'
  | 'es2015'
  | 'es2016'
  | 'es2017'
  | 'es2018'
  | 'es2019'
  | 'es2020'
  | 'es2021'
  | 'es2022'
  | 'es2023';
export type LatestEcmaVersions = 'es2024' | 'esnext';
export type EcmaScriptVersion = FixedEcmaVersions | LatestEcmaVersions;

export type RsbuildConfigOutputTarget = NonNullable<
  RsbuildConfig['output']
>['target'];

export type Syntax =
  // ECMAScript versions as an common used addition to browserslist query
  | EcmaScriptVersion
  // Support inline browserslist query, like defined in package.json
  | string[];

export type Dts =
  | (Pick<PluginDtsOptions, 'bundle' | 'distPath' | 'abortOnError'> & {
      autoExtension?: boolean;
    })
  | false;

export type AutoExternal =
  | boolean
  | {
      dependencies?: boolean;
      devDependencies?: boolean;
      peerDependencies?: boolean;
    };

export type BannerAndFooter = {
  js?: string;
  css?: string;
  dts?: string;
};

export type Redirect = {
  // TODO: support other redirects
  // alias?: boolean;
  style?: boolean;
  // asset?: boolean;
  // autoExtension?: boolean;
};

export interface LibConfig extends RsbuildConfig {
  bundle?: boolean;
  format?: Format;
  autoExtension?: boolean;
  autoExternal?: AutoExternal;
  redirect?: Redirect;
  /** Support esX and browserslist query */
  syntax?: Syntax;
  externalHelpers?: boolean;
  banner?: BannerAndFooter;
  footer?: BannerAndFooter;
  dts?: Dts;
}

export interface RslibConfig extends RsbuildConfig {
  lib: LibConfig[];
}

export type ConfigParams = {
  env: string;
  command: string;
  envMode?: string;
};

export type RslibConfigSyncFn = (env: ConfigParams) => RslibConfig;
export type RslibConfigAsyncFn = (env: ConfigParams) => Promise<RslibConfig>;
export type RslibConfigExport =
  | RslibConfig
  | RslibConfigSyncFn
  | RslibConfigAsyncFn;
