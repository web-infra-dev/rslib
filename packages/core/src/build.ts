import type { RsbuildInstance } from '@rsbuild/core';
import type { BuildOptions } from './cli/commands';
import { initRsbuild } from './config';
import type { RslibConfig } from './types/config';

export async function build(
  config: RslibConfig,
  options?: BuildOptions,
): Promise<RsbuildInstance> {
  const rsbuildInstance = await initRsbuild(config);

  await rsbuildInstance.build({
    mode: 'production',
    watch: options?.watch,
  });

  return rsbuildInstance;
}
