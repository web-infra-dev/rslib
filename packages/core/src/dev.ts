import { initDevMFRsbuild } from './config';

import type { RsbuildInstance } from '@rsbuild/core';
import type { RslibConfig } from './types/config';

export async function dev(config: RslibConfig): Promise<RsbuildInstance> {
  const rsbuildInstance = await initDevMFRsbuild(config);

  await rsbuildInstance.startDevServer();

  return rsbuildInstance;
}
