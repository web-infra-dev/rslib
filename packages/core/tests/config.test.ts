import { join } from 'node:path';
import { describe, expect, test, vi } from 'vitest';
import { composeCreateRsbuildConfig, loadConfig } from '../src/config';
import type { RslibConfig } from '../src/types/config';

vi.mock('rslog');

describe('Should load config file correctly', () => {
  test('Load config.js in cjs project', async () => {
    const fixtureDir = join(__dirname, 'fixtures/config/cjs');
    const configFilePath = join(fixtureDir, 'rslib.config.js');
    const config = await loadConfig({ path: configFilePath });
    expect(config).toEqual({
      lib: [],
      source: {
        entry: {
          index: './foo/index.js',
        },
      },
      _privateMeta: {
        configFilePath,
      },
    });
  });

  test('Load config.mjs in cjs project', async () => {
    const fixtureDir = join(__dirname, 'fixtures/config/cjs');
    const configFilePath = join(fixtureDir, 'rslib.config.mjs');
    const config = await loadConfig({ path: configFilePath });
    expect(config).toEqual({
      lib: [],
      source: {
        entry: {
          index: './foo/index.js',
        },
      },
      _privateMeta: {
        configFilePath,
      },
    });
  });

  test('Load config.ts in cjs project', async () => {
    const fixtureDir = join(__dirname, 'fixtures/config/cjs');
    const configFilePath = join(fixtureDir, 'rslib.config.ts');
    const config = await loadConfig({ path: configFilePath });
    expect(config).toEqual({
      lib: [],
      source: {
        entry: {
          index: './foo/index.ts',
        },
      },
      _privateMeta: {
        configFilePath,
      },
    });
  });

  test('Load config.cjs with defineConfig in cjs project', async () => {
    const fixtureDir = join(__dirname, 'fixtures/config/cjs');
    const configFilePath = join(fixtureDir, 'rslib.config.cjs');
    const config = await loadConfig({ path: configFilePath });
    expect(config).toEqual({
      lib: [],
      source: {
        entry: {
          index: './foo/index.js',
        },
      },
      _privateMeta: {
        configFilePath,
      },
    });
  });

  test('Load config.js in esm project', async () => {
    const fixtureDir = join(__dirname, 'fixtures/config/esm');
    const configFilePath = join(fixtureDir, 'rslib.config.js');
    const config = await loadConfig({ path: configFilePath });
    expect(config).toEqual({
      lib: [],
      source: {
        entry: {
          index: './foo/index.js',
        },
      },
      _privateMeta: {
        configFilePath,
      },
    });
  });

  test('Load config.cjs in esm project', async () => {
    const fixtureDir = join(__dirname, 'fixtures/config/esm');
    const configFilePath = join(fixtureDir, 'rslib.config.cjs');
    const config = await loadConfig({ path: configFilePath });
    expect(config).toEqual({
      lib: [],
      source: {
        entry: {
          index: './foo/index.js',
        },
      },
      _privateMeta: {
        configFilePath,
      },
    });
  });

  test('Load config.ts in esm project', async () => {
    const fixtureDir = join(__dirname, 'fixtures/config/esm');
    const configFilePath = join(fixtureDir, 'rslib.config.ts');
    const config = await loadConfig({ path: configFilePath });
    expect(config).toEqual({
      lib: [],
      source: {
        entry: {
          index: './foo/index.ts',
        },
      },
      _privateMeta: {
        configFilePath,
      },
    });
  });

  test('Load config.mjs with defineConfig in esm project', async () => {
    const fixtureDir = join(__dirname, 'fixtures/config/esm');
    const configFilePath = join(fixtureDir, 'rslib.config.mjs');
    const config = await loadConfig({ path: configFilePath });
    expect(config).toMatchObject({
      lib: [],
      source: {
        entry: {
          index: './foo/index.js',
        },
      },
      _privateMeta: {
        configFilePath,
      },
    });
  });
});

describe('Should compose create Rsbuild config correctly', () => {
  test('Merge Rsbuild config', async () => {
    const rslibConfig: RslibConfig = {
      lib: [
        {
          format: 'esm',
          source: {
            alias: {
              foo: 'foo/esm',
            },
            preEntry: './b.js',
          },
        },
        {
          format: 'cjs',
          source: {
            alias: {
              bar: 'bar/cjs',
            },
            preEntry: ['./c.js', './d.js'],
          },
        },
        {
          format: 'umd',
        },
      ],
      source: {
        alias: {
          foo: 'foo',
          bar: 'bar',
        },
        preEntry: './a.js',
      },
      output: {
        filenameHash: false,
      },
      tools: {
        rspack: {
          resolve: {
            extensionAlias: {
              '.js': ['.ts', '.tsx'],
            },
          },
        },
      },
    };
    const composedRsbuildConfig = await composeCreateRsbuildConfig(
      rslibConfig,
      process.cwd(),
    );
    expect(composedRsbuildConfig).toMatchSnapshot();
  });
});

describe('syntax', () => {
  test('`syntax` default value', async () => {
    const rslibConfig: RslibConfig = {
      lib: [
        {
          format: 'esm',
        },
      ],
    };

    const composedRsbuildConfig = await composeCreateRsbuildConfig(
      rslibConfig,
      process.cwd(),
    );

    expect(
      composedRsbuildConfig[0].config.output?.overrideBrowserslist,
    ).toMatchInlineSnapshot(`
      [
        "last 1 node versions",
        "last 1 Chrome versions",
        "last 1 Firefox versions",
        "last 1 Edge versions",
        "last 1 Safari versions",
        "last 1 ios_saf versions",
        "not dead",
      ]
    `);
  });

  test('`syntax` default value should determined by target `web`', async () => {
    const rslibConfig: RslibConfig = {
      lib: [
        {
          format: 'esm',
        },
      ],
      output: {
        target: 'web',
      },
    };

    const composedRsbuildConfig = await composeCreateRsbuildConfig(
      rslibConfig,
      process.cwd(),
    );

    expect(
      composedRsbuildConfig[0].config.output?.overrideBrowserslist,
    ).toMatchInlineSnapshot(`
      [
        "last 1 Chrome versions",
        "last 1 Firefox versions",
        "last 1 Edge versions",
        "last 1 Safari versions",
        "last 1 ios_saf versions",
        "not dead",
      ]
    `);
  });

  test('`syntax` default value should determined by target `node`', async () => {
    const rslibConfig: RslibConfig = {
      lib: [
        {
          format: 'esm',
        },
      ],
      output: {
        target: 'node',
      },
    };

    const composedRsbuildConfig = await composeCreateRsbuildConfig(
      rslibConfig,
      process.cwd(),
    );

    expect(
      composedRsbuildConfig[0].config.output?.overrideBrowserslist,
    ).toMatchInlineSnapshot(`
      [
        "last 1 node versions",
      ]
    `);
  });

  test('`syntax` could accept `esX` and transform to browserslist', async () => {
    const rslibConfig: RslibConfig = {
      lib: [
        {
          syntax: 'es2015',
          format: 'esm',
        },
      ],
    };

    const composedRsbuildConfig = await composeCreateRsbuildConfig(
      rslibConfig,
      process.cwd(),
    );

    expect(
      composedRsbuildConfig[0].config.output?.overrideBrowserslist,
    ).toMatchInlineSnapshot(`
      [
        "Chrome >= 63.0.0",
        "Edge >= 79.0.0",
        "Firefox >= 67.0.0",
        "iOS >= 13.0.0",
        "Node >= 10.0.0",
        "Opera >= 50.0.0",
        "Safari >= 13.0.0",
      ]
    `);
  });
});

describe('minify', () => {
  test('`minify` default value', async () => {
    const rslibConfig: RslibConfig = {
      lib: [
        {
          format: 'esm',
        },
      ],
    };

    const composedRsbuildConfig = await composeCreateRsbuildConfig(
      rslibConfig,
      process.cwd(),
    );

    expect(
      composedRsbuildConfig[0].config.output?.minify,
    ).toMatchInlineSnapshot(`
      {
        "css": false,
        "js": true,
        "jsOptions": {
          "minimizerOptions": {
            "compress": {
              "dead_code": true,
              "defaults": false,
              "toplevel": true,
              "unused": true,
            },
            "format": {
              "comments": "all",
            },
            "mangle": false,
            "minify": false,
          },
        },
      }
    `);
  });

  test('`minify` is configured by user', async () => {
    const rslibConfig: RslibConfig = {
      lib: [
        {
          format: 'esm',
          output: {
            minify: false,
          },
        },
        {
          format: 'esm',
          output: {
            minify: true,
          },
        },
        {
          format: 'esm',
          output: {
            minify: {
              js: false,
              css: true,
            },
          },
        },
      ],
      output: {
        target: 'web',
      },
    };

    const composedRsbuildConfig = await composeCreateRsbuildConfig(
      rslibConfig,
      process.cwd(),
    );

    expect(
      composedRsbuildConfig[0].config.output?.minify,
    ).toMatchInlineSnapshot('false');

    expect(
      composedRsbuildConfig[1].config.output?.minify,
    ).toMatchInlineSnapshot('true');

    expect(
      composedRsbuildConfig[2].config.output?.minify,
    ).toMatchInlineSnapshot(`
      {
        "css": true,
        "js": false,
      }
    `);
  });
});
