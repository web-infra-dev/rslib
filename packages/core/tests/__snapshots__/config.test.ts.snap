// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should compose create Rsbuild config correctly > Merge Rsbuild config 1`] = `
[
  {
    "config": {
      "dev": {
        "progressBar": false,
      },
      "mode": "production",
      "output": {
        "distPath": {
          "js": "./",
        },
        "externals": [
          [Function],
        ],
        "filename": {
          "js": "[name].js",
        },
        "filenameHash": false,
        "minify": {
          "css": false,
          "js": true,
          "jsOptions": {
            "minimizerOptions": {
              "compress": {
                "dead_code": true,
                "defaults": false,
                "toplevel": true,
                "unused": true,
              },
              "format": {
                "comments": "all",
              },
              "mangle": false,
              "minify": false,
            },
          },
        },
        "overrideBrowserslist": [
          "last 1 node versions",
          "last 1 Chrome versions",
          "last 1 Firefox versions",
          "last 1 Edge versions",
          "last 1 Safari versions",
          "last 1 ios_saf versions",
          "not dead",
        ],
      },
      "source": {
        "alias": {
          "bar": "bar",
          "foo": "foo/esm",
        },
        "entry": {},
        "preEntry": "./b.js",
      },
      "tools": {
        "htmlPlugin": false,
        "rspack": [
          {
            "experiments": {
              "rspackFuture": {
                "bundlerInfo": {
                  "force": false,
                },
              },
            },
            "optimization": {
              "moduleIds": "named",
              "nodeEnv": false,
            },
            "resolve": {
              "extensionAlias": {
                ".cjs": [
                  ".cts",
                  ".cjs",
                ],
                ".js": [
                  ".ts",
                  ".tsx",
                  ".js",
                  ".jsx",
                ],
                ".jsx": [
                  ".tsx",
                  ".jsx",
                ],
                ".mjs": [
                  ".mts",
                  ".mjs",
                ],
              },
            },
          },
          {
            "experiments": {
              "outputModule": true,
            },
            "externalsType": "module-import",
            "module": {
              "parser": {
                "javascript": {
                  "importDynamic": false,
                  "importMeta": false,
                  "requireAsExpression": false,
                  "requireDynamic": false,
                  "requireResolve": false,
                },
              },
            },
            "optimization": {
              "concatenateModules": true,
              "sideEffects": "flag",
            },
            "output": {
              "chunkFormat": "module",
              "chunkLoading": "import",
              "library": {
                "type": "modern-module",
              },
              "module": true,
              "wasmLoading": "fetch",
              "workerChunkLoading": "import",
            },
          },
          [Function],
          {
            "target": [
              "web",
            ],
          },
          {
            "resolve": {
              "extensionAlias": {
                ".js": [
                  ".ts",
                  ".tsx",
                ],
              },
            },
          },
        ],
        "swc": {
          "jsc": {
            "externalHelpers": false,
          },
        },
      },
    },
    "format": "esm",
  },
  {
    "config": {
      "dev": {
        "progressBar": false,
      },
      "mode": "production",
      "output": {
        "distPath": {
          "js": "./",
        },
        "filename": {
          "js": "[name].js",
        },
        "filenameHash": false,
        "minify": {
          "css": false,
          "js": true,
          "jsOptions": {
            "minimizerOptions": {
              "compress": {
                "dead_code": true,
                "defaults": false,
                "toplevel": true,
                "unused": true,
              },
              "format": {
                "comments": "all",
              },
              "mangle": false,
              "minify": false,
            },
          },
        },
        "overrideBrowserslist": [
          "last 1 node versions",
          "last 1 Chrome versions",
          "last 1 Firefox versions",
          "last 1 Edge versions",
          "last 1 Safari versions",
          "last 1 ios_saf versions",
          "not dead",
        ],
      },
      "plugins": [
        {
          "name": "rsbuild-plugin-cjs-shim",
          "setup": [Function],
        },
      ],
      "source": {
        "alias": {
          "bar": "bar/cjs",
          "foo": "foo",
        },
        "entry": {},
        "preEntry": [
          "./a.js",
          "./c.js",
          "./d.js",
        ],
      },
      "tools": {
        "htmlPlugin": false,
        "rspack": [
          {
            "experiments": {
              "rspackFuture": {
                "bundlerInfo": {
                  "force": false,
                },
              },
            },
            "optimization": {
              "moduleIds": "named",
              "nodeEnv": false,
            },
            "resolve": {
              "extensionAlias": {
                ".cjs": [
                  ".cts",
                  ".cjs",
                ],
                ".js": [
                  ".ts",
                  ".tsx",
                  ".js",
                  ".jsx",
                ],
                ".jsx": [
                  ".tsx",
                  ".jsx",
                ],
                ".mjs": [
                  ".mts",
                  ".mjs",
                ],
              },
            },
          },
          {
            "externalsType": "commonjs",
            "module": {
              "parser": {
                "javascript": {
                  "importDynamic": false,
                  "importMeta": false,
                  "requireAsExpression": false,
                  "requireDynamic": false,
                  "requireResolve": false,
                },
              },
            },
            "output": {
              "chunkFormat": "commonjs",
              "chunkLoading": "require",
              "iife": false,
              "library": {
                "type": "commonjs",
              },
              "wasmLoading": "async-node",
              "workerChunkLoading": "async-node",
            },
          },
          [Function],
          {
            "target": [
              "web",
            ],
          },
          {
            "resolve": {
              "extensionAlias": {
                ".js": [
                  ".ts",
                  ".tsx",
                ],
              },
            },
          },
        ],
        "swc": {
          "jsc": {
            "externalHelpers": false,
          },
        },
      },
    },
    "format": "cjs",
  },
  {
    "config": {
      "dev": {
        "progressBar": false,
      },
      "mode": "production",
      "output": {
        "distPath": {
          "js": "./",
        },
        "filename": {
          "js": "[name].js",
        },
        "filenameHash": false,
        "minify": {
          "css": false,
          "js": true,
          "jsOptions": {
            "minimizerOptions": {
              "compress": {
                "dead_code": true,
                "defaults": false,
                "toplevel": true,
                "unused": true,
              },
              "format": {
                "comments": "all",
              },
              "mangle": false,
              "minify": false,
            },
          },
        },
        "overrideBrowserslist": [
          "last 1 node versions",
          "last 1 Chrome versions",
          "last 1 Firefox versions",
          "last 1 Edge versions",
          "last 1 Safari versions",
          "last 1 ios_saf versions",
          "not dead",
        ],
      },
      "source": {
        "alias": {
          "bar": "bar",
          "foo": "foo",
        },
        "entry": {},
        "preEntry": "./a.js",
      },
      "tools": {
        "htmlPlugin": false,
        "rspack": [
          {
            "experiments": {
              "rspackFuture": {
                "bundlerInfo": {
                  "force": false,
                },
              },
            },
            "optimization": {
              "moduleIds": "named",
              "nodeEnv": false,
            },
            "resolve": {
              "extensionAlias": {
                ".cjs": [
                  ".cts",
                  ".cjs",
                ],
                ".js": [
                  ".ts",
                  ".tsx",
                  ".js",
                  ".jsx",
                ],
                ".jsx": [
                  ".tsx",
                  ".jsx",
                ],
                ".mjs": [
                  ".mts",
                  ".mjs",
                ],
              },
            },
          },
          {
            "externalsType": "umd",
            "module": {
              "parser": {
                "javascript": {
                  "importMeta": false,
                },
              },
            },
            "output": {
              "library": {
                "type": "umd",
              },
            },
          },
          [Function],
          {
            "target": [
              "web",
            ],
          },
          {
            "resolve": {
              "extensionAlias": {
                ".js": [
                  ".ts",
                  ".tsx",
                ],
              },
            },
          },
        ],
        "swc": {
          "jsc": {
            "externalHelpers": false,
          },
        },
      },
    },
    "format": "umd",
  },
]
`;
